### ISSUE TITLE
Chakra-ui inside shadow dom

### ISSUE THREAD

**Issue by pserrer1** [chakra-ui/chakra-ui#2145]
# Feature request

Provide a way to use chakra-ui inside a shadow dom

## Is your feature request related to a problem? Please describe.

Currently you can't use chakra-ui inside a shadow dom as emotion adds all style rules to the main page header element and not the shadow doms root element. As the shadow dom is isolated from the outside dom the rules aren't available.

## Describe the solution you'd like

It would be cool if chakra detects that it's rendered inside a shadow dom and then automatically adds the style rules to the shadow root. Otherwise it would also be an option to specify to styles root container.

**Comment by ljosberinn** [comment]
Is this something emotion supports? I'm afraid we're bound to what they support in this case.

**Comment by pserrer1** [comment]
Emotion supports defining a container with the low level function createCache https://emotion.sh/docs/@emotion/cache#container

**Comment by pserrer1** [comment]
I solved my issue with the following workaround using `CacheProvider`
```
const cache = createCache({
   ...
   container: someDomElementInShadowDom
});
```

and later

```
return (
  <CacheProvider value={cache}>
    ...
    <Button>Foo</Button>
    ...
  </CacheProvider>
);
```

One important thing about this solution is, that `@chakra/system` and your own package uses the same `@emotion/core` version. Otherwise they wont play together.

**Comment by ljosberinn** [comment]
Interesting, and fairly simple too actually. I'm honestly curious about the use case however, would you mind expanding a bit on this, maybe even with a codesandbox?

**Comment by pserrer1** [comment]
Our use case was inside a single-spa application where some applications are isolated in a shadow dom. As an example for chakra-ui inside a shadow dom you could have a look into https://codesandbox.io/s/intelligent-einstein-cywkq
Thats not exactly what we did, but could maybe a good starting point.

**Comment by ljosberinn** [comment]
Nice, haven't heard of isolating _entire apps_ in shadow dom yet. You used v0.8 in the sandbox, are you aware of v1? :)

**Comment by baptisteArnaud** [comment]
Hi @pserrer1 ,

Thank you for these insights. How do you provide a custom chakra theme to your Shadow DOM elements?

Here is your forked sandbox showing the problem: https://codesandbox.io/s/falling-leftpad-uw6it?file=/src/ShadowChakra.js

**Comment by cameronbraid** [comment]
Doesn't seem to work with v1

https://codesandbox.io/s/chakra-ui-and-react-shadow-broken-o1eqr

**Comment by k4m4** [comment]
@cameronbraid Were you able to figure out how to get `react-shadow` to work with chakra-ui `^1.0`?

**Comment by cameronbraid** [comment]
@k4m4 nope - but I didn't try very hard either, had other priorities to move onto

**Comment by k4m4** [comment]
@ljosberinn Any thoughts on why the code in @cameronbraid's CodeSandbox ([a few comments above](https://github.com/chakra-ui/chakra-ui/issues/2145#issuecomment-826100439)) doesn't work?

**Comment by ljosberinn** [comment]
it looks like css variables aren't set to the corresponding shadow root with `:root` 🤔 everything unrelated to variables defined within the theme works, but variables are applied to nothing instead

**Comment by k4m4** [comment]
@ljosberinn I see - thanks for the quick reply! Is there any temporary fix that you can think of, other then rolling back to an old version of Chakra-UI?

**Comment by pserrer1** [comment]
You can work around the issue with an stylis plugin in the emotion cache, replacing all `:root` selectors, but for me it seems, that should be something done inside chakra-ui

**Comment by ljosberinn** [comment]
@k4m4 I cant whip up a test sandbox right now but it should be possible to temp fix this yourself on latest version by orchestrating the providers the `ChakraProvider` abstracts away for you yourself.

Instead of using `ChakraProvider`, import the providers its rendering and those you need yourself. [Here's its implementation](https://github.com/chakra-ui/chakra-ui/blob/main/packages/react/src/chakra-provider.tsx#L73).

Then, to the `ThemeProvider`, pass the `cssVarsRoot` property with `:host,:root`.

**Comment by k4m4** [comment]
@ljosberinn Thanks a lot! I'll give it a shot soon and let you know.

**Comment by k4m4** [comment]
@ljosberinn That did it, for now. Thanks again!

**Comment by avalanche1** [comment]
@segunadebayo I still don't get - what is the code to have Chakra in my react app that is in shadow dom?
I'm using chakra-ui `^1.8.8`

**Comment by EmilEriksen** [comment]
Here's an example of how to use Chakra and Emotion in shadow DOM. This way styles are added in the shadow DOM. Portals are still rendered in the body though.

```jsx
const rootEl = document.createElement('div')

const shadowRoot = this.attachShadow({ mode: 'open' })

shadowRoot.appendChild(rootEl)

const cache = createCache({
    key: 'css',
    container: shadowRoot,
})

const root = createRoot(rootEl)
root.render(
    <CacheProvider value={cache}>
        <ChakraProvider>
            <App/>
        </ChakraProvider>
    </CacheProvider>
)
```

**Comment by and3rson** [comment]
Is it possible to globally specify a different container for portals? I'm using @EmilEriksen's solution, but things like drawers have broken styles since they are rendered outside of shadow DOM.

**Comment by wearsshoes** [comment]
having same problem as @and3rson ; I'd like to put tooltips on some of the elements of my chrome extension's onscreen widget, but the tooltips render outside of the shadow DOM and so don't get positioned properly or rendered correctly.

**Comment by and3rson** [comment]
Bumping this issue since we've once again reached the point where we have to use shadow DOM.

**Comment by and3rson** [comment]
Created #8566 with a proposed fix.

**Comment by and3rson** [comment]
I found another solution: creating a parent portal.

It turns out that Chakra will respect parent portals, so we can wrap entire app into portal and force all Chakra's portals to be appended into it, thus encapsulating everything within a Shadow DOM:

```tsx
const WidgetWrapper = ({ root }: any) => {
  // This component puts entire app inside a Chakra Portal.
  //
  // Why? Because Chakra uses portals to render popups/dialogs/tooltips.
  // Chakra's default portal container is `document.body`, which is outside the shadow DOM, and we cannot change it.
  //
  // This means that popups/dialogs/tooltips will not be rendered inside the shadow DOM, and will be rendered outside the shadow DOM.
  //
  // However, by default Chakra will respect its parent portal. So if we put the entire app inside a Chakra portal,
  // all child portals will be rendered inside the parent portal (that we create here), which is inside the shadow DOM.
  const cache = createCache({ key: 'chakra-css', container: root });
  const rootRef = useRef<HTMLDivElement>(root);
  return (
    <Portal containerRef={rootRef}>
      <CacheProvider value={cache}>
        <ChakraProvider>
          <Widget />
        </ChakraProvider>
      </CacheProvider>
    </Portal>
  );
};

(window as any).initWidget = ({ el }: any) => {
  const shadowRoot = el.attachShadow({ mode: 'open' });
  const reactRoot = ReactDOM.createRoot(shadowRoot);
  reactRoot.render(
    <WidgetWrapper root={shadowRoot} />
  );
  return () => reactRoot.unmount();
}
```

**Comment by kstulgys** [comment]
> I found another solution: creating a parent portal.
> 
> It turns out that Chakra will respect parent portals, so we can wrap entire app into portal and force all Chakra's portals to be appended into it, thus encapsulating everything within a Shadow DOM:
> 
> > interface PortalProps {
> > /**
> > * The `ref` to the component where the portal will be attached to.
> > */
> > containerRef?: React.RefObject<HTMLElement | null>;
> 
> TL;DR:
> 
> * Create Chakra `Portal` inside shadow DOM.
> * Add `ChakraProvider` and all its children into this portal.
> 
> Working example:
> 
> ```tsx
> const WidgetWrapper = ({ root }: any) => {
>   // This component puts entire app inside a Chakra Portal.
>   //
>   // Why? Because Chakra uses portals to render popups/dialogs/tooltips.
>   // Chakra's default portal container is `document.body`,
>   // which is outside the shadow DOM, and we cannot change it.
>   //
>   // This means that popups/dialogs/tooltips will not be rendered inside the shadow DOM,
>   // and will be rendered outside the shadow DOM.
>   //
>   // However, by default Chakra will respect its parent portal.
>   // So if we put the entire app inside a Chakra portal, all child portals will be rendered
>   // inside the parent portal (that we create here), which is inside the shadow DOM.
>   const cache = createCache({ key: 'chakra-css', container: root });
>   const rootRef = useRef<HTMLDivElement>(root);
>   return (
>     <Portal containerRef={rootRef}>
>       <CacheProvider value={cache}>
>         <ChakraProvider>
>           <Widget />
>         </ChakraProvider>
>       </CacheProvider>
>     </Portal>
>   );
> };
> 
> (window as any).initWidget = ({ el }: any) => {
>   const shadowRoot = el.attachShadow({ mode: 'open' });
>   const reactRoot = ReactDOM.createRoot(shadowRoot);
>   reactRoot.render(
>     <WidgetWrapper root={shadowRoot} />
>   );
>   return () => reactRoot.unmount();
> }
> ```

@and3rson you are legend! I was fighting chakra and shadow dom for 2 days. Here is my code:

```
const WidgetWrapper = ({ root }: { root: HTMLBodyElement }) => {
  const cache = createCache({ key: "c", container: root });
  const rootRef = React.useRef<HTMLBodyElement>(root);
  return (
    <Portal containerRef={rootRef}>
      <CacheProvider value={cache}>
        <ChakraProvider theme={theme}>
          <App />
        </ChakraProvider>
      </CacheProvider>
    </Portal>
  );
};

const rootElement = (() => {
  const div = document.createElement("div");
  div.id = "snip";
  const body = document.createElement("body");
  document.body.appendChild(div);
  const shadowDom = div.attachShadow({ mode: "open" });
  shadowDom.appendChild(body);
  return body;
})();

const initSnapCart = () => {
  let reactRoot: ReactDOM.Root;
  setTimeout(() => {
    reactRoot = ReactDOM.createRoot(rootElement);
    reactRoot.render(<WidgetWrapper root={rootElement} />);
  }, 3000);
  return () => reactRoot?.unmount?.();
};

document.addEventListener("DOMContentLoaded", initSnapCart);
```
A few changes:
i noticed if I use body tag chakra now gets font family
I wait for 3 seconds to load because if I execute script I get hydration error in nextjs15

**Comment by lucksp** [comment]
This is still a thing...in case wondering
```
"@chakra-ui/react": "^3.27.0",
    "@emotion/cache": "^11.14.0",
    "@emotion/react": "^11.14.0",
```

solving it by:

system config:
```js
const config = defineConfig({
  cssVarsRoot: varRoot,
  conditions: {
    light: `:host([data-theme=light]) &, ${varRoot} &`,
    dark: `:host([data-theme=dark]) &`
  },
  preflight: { scope: varRoot },
  globalCss: {
    [varRoot]: defaultConfig.globalCss?.html ?? {}
  },
  theme: {
   ...
  }
});
```
and then Providers:

```js
export const Providers = ({
  children,
  colorMode = 'light'
}: ProvidersProps) => {
  const [shadow, setShadow] = useState<HTMLElement | null>(null);
  const portalRef = useRef<HTMLDivElement>(null);

  const cache = useMemo(() => {
    if (!shadow?.shadowRoot) return null;
    return createCache({
      key: 'identafly-widget',
      container: shadow.shadowRoot
    });
  }, [shadow]);

  // Apply data-theme to the shadow host element directly — never touching document.documentElement
  useEffect(() => {
    if (!shadow) return;
    shadow.setAttribute('data-theme', colorMode);
    shadow.style.setProperty('color-scheme', colorMode);
  }, [shadow, colorMode]);

  return (
    <root.div ref={setShadow}>
      {shadow && cache && (
        <EnvironmentProvider value={() => shadow.shadowRoot ?? document}>
          <CacheProvider value={cache}>
            <ChakraProvider value={system}>
              {/* Portal mount point — must render before children so the ref
                  is populated before any dialog/tooltip tries to use it */}
              <div ref={portalRef} />
              <WidgetPortalContext.Provider value={portalRef}>
                {children}
              </WidgetPortalContext.Provider>
            </ChakraProvider>
          </CacheProvider>
        </EnvironmentProvider>
      )}
    </root.div>
  );
};

```
